File: backend/scripts/bulk_import.py
import argparse
import csv
import os
import sys
import uuid

import bcrypt
import psycopg2
from dotenv import load_dotenv


REQUIRED_USER_FIELDS = ["alias", "first_name", "last_name", "role", "pin"]
VALID_ROLES = {"student", "mentor", "coach"}


def load_env():
    load_dotenv()


def get_connection():
    conn = psycopg2.connect(
        host=os.getenv("DB_HOST", "localhost"),
        port=int(os.getenv("DB_PORT", "5432")),
        dbname=os.getenv("DB_NAME", "robotics_attendance"),
        user=os.getenv("DB_USER", "postgres"),
        password=os.getenv("DB_PASSWORD"),
        connect_timeout=5,
    )
    return conn


def hash_pin(pin: str) -> str:
    salt = bcrypt.gensalt()
    return bcrypt.hashpw(pin.encode("utf-8"), salt).decode("utf-8")


def validate_row(row: dict):
    for field in REQUIRED_USER_FIELDS:
        if not row.get(field):
            raise ValueError(f"Missing required field: {field}")
    role = row.get("role", "").lower()
    if role not in VALID_ROLES:
        raise ValueError(f"Invalid role '{role}'. Must be one of {sorted(VALID_ROLES)}")


def upsert_user(cur, row: dict):
    alias = row["alias"].strip()
    validate_row(row)

    cur.execute("SELECT id FROM users WHERE alias = %s", (alias,))
    existing = cur.fetchone()

    pin_hash = hash_pin(row["pin"].strip()) if row.get("pin") else None
    user_id = existing[0] if existing else str(uuid.uuid4())

    if existing:
        # Update existing user
        cur.execute(
            """
            UPDATE users
            SET first_name = %s,
                middle_name = %s,
                last_name = %s,
                role = %s,
                is_active = true,
                updated_at = CURRENT_TIMESTAMP,
                pin_hash = COALESCE(%s, pin_hash)
            WHERE id = %s
            RETURNING id
            """,
            (
                row.get("first_name"),
                row.get("middle_name") or None,
                row.get("last_name"),
                row.get("role").lower(),
                pin_hash,
                user_id,
            ),
        )
        return user_id, "updated"
    else:
        if not pin_hash:
            raise ValueError("Pin is required for new users")
        cur.execute(
            """
            INSERT INTO users (id, first_name, middle_name, last_name, alias, role, pin_hash, is_active)
            VALUES (%s, %s, %s, %s, %s, %s, %s, true)
            RETURNING id
            """,
            (
                user_id,
                row.get("first_name"),
                row.get("middle_name") or None,
                row.get("last_name"),
                alias,
                row.get("role").lower(),
                pin_hash,
            ),
        )
        return user_id, "inserted"


def upsert_parent_contact(cur, user_id: str, row: dict):
    name = (row.get("parent_name") or "").strip()
    phone = (row.get("parent_phone") or "").strip()
    relationship = (row.get("parent_relationship") or "").strip() or None

    if not name or not phone:
        return None, "skipped"

    cur.execute(
        "SELECT id FROM parent_contacts WHERE user_id = %s AND phone_number = %s",
        (user_id, phone),
    )
    existing = cur.fetchone()

    if existing:
        cur.execute(
            """
            UPDATE parent_contacts
            SET name = %s,
                relationship = %s,
                updated_at = CURRENT_TIMESTAMP
            WHERE id = %s
            RETURNING id
            """,
            (name, relationship, existing[0]),
        )
        return existing[0], "updated"
    else:
        contact_id = str(uuid.uuid4())
        cur.execute(
            """
            INSERT INTO parent_contacts (id, user_id, name, phone_number, relationship)
            VALUES (%s, %s, %s, %s, %s)
            RETURNING id
            """,
            (contact_id, user_id, name, phone, relationship),
        )
        return contact_id, "inserted"


def process_csv(path: str, dry_run: bool = False):
    load_env()
    conn = get_connection()
    summary = {"users_inserted": 0, "users_updated": 0, "contacts_inserted": 0, "contacts_updated": 0, "contacts_skipped": 0, "rows_processed": 0, "rows_failed": 0}

    with conn:
        with conn.cursor() as cur:
            with open(path, newline="", encoding="utf-8") as csvfile:
                reader = csv.DictReader(csvfile)
                for idx, row in enumerate(reader, start=1):
                    try:
                        user_id, status = upsert_user(cur, row)
                        summary[f"users_{status}"] += 1
                        contact_id, c_status = upsert_parent_contact(cur, user_id, row)
                        if c_status == "inserted":
                            summary["contacts_inserted"] += 1
                        elif c_status == "updated":
                            summary["contacts_updated"] += 1
                        else:
                            summary["contacts_skipped"] += 1
                        summary["rows_processed"] += 1
                    except Exception as exc:
                        summary["rows_failed"] += 1
                        print(f"Row {idx} failed: {exc}", file=sys.stderr)
            if dry_run:
                conn.rollback()
                print("Dry run complete; no changes committed.")
            else:
                conn.commit()

    return summary


def main():
    parser = argparse.ArgumentParser(description="Bulk import users and parent contacts from CSV")
    parser.add_argument("csv_path", help="Path to CSV file")
    parser.add_argument("--dry-run", action="store_true", help="Validate without committing changes")
    args = parser.parse_args()

    summary = process_csv(args.csv_path, dry_run=args.dry_run)
    print("\nImport summary:")
    for key, value in summary.items():
        print(f"  {key}: {value}")


if __name__ == "__main__":
    main()

File: backend/scripts/bulk_import.js
const fs = require('fs');
const path = require('path');
const { parse } = require('csv-parse/sync');
const { v4: uuidv4 } = require('uuid');
const bcrypt = require('bcryptjs');
require('dotenv').config({ path: path.join(__dirname, '../.env') });
const { pool } = require('../config/database');

const REQUIRED_USER_FIELDS = ['alias', 'first_name', 'last_name', 'role', 'pin'];
const VALID_ROLES = new Set(['student', 'mentor', 'coach']);

function validateRow(row) {
  for (const field of REQUIRED_USER_FIELDS) {
    if (!row[field] || !row[field].trim()) {
      throw new Error(`Missing required field: ${field}`);
    }
  }
  const role = (row.role || '').toLowerCase();
  if (!VALID_ROLES.has(role)) {
    throw new Error(`Invalid role '${role}'. Must be one of: ${Array.from(VALID_ROLES).sort().join(', ')}`);
  }
}

async function upsertUser(client, row) {
  const alias = row.alias.trim();
  validateRow(row);

  // Check if user exists
  const existingResult = await client.query(
    'SELECT id FROM users WHERE alias = $1',
    [alias]
  );
  const existing = existingResult.rows[0];

  const pinHash = row.pin ? await bcrypt.hash(row.pin.trim(), 10) : null;
  const userId = existing ? existing.id : uuidv4();

  if (existing) {
    // Update existing user
    await client.query(
      `UPDATE users
       SET first_name = $1,
           middle_name = $2,
           last_name = $3,
           role = $4,
           is_active = true,
           updated_at = CURRENT_TIMESTAMP,
           pin_hash = COALESCE($5, pin_hash)
       WHERE id = $6`,
      [
        row.first_name,
        row.middle_name || null,
        row.last_name,
        row.role.toLowerCase(),
        pinHash,
        userId
      ]
    );
    return { userId, status: 'updated' };
  } else {
    // Insert new user
    if (!pinHash) {
      throw new Error('Pin is required for new users');
    }
    await client.query(
      `INSERT INTO users (id, first_name, middle_name, last_name, alias, role, pin_hash, is_active)
       VALUES ($1, $2, $3, $4, $5, $6, $7, true)`,
      [
        userId,
        row.first_name,
        row.middle_name || null,
        row.last_name,
        alias,
        row.role.toLowerCase(),
        pinHash
      ]
    );
    return { userId, status: 'inserted' };
  }
}

async function upsertParentContact(client, userId, row) {
  const name = (row.parent_name || '').trim();
  const phone = (row.parent_phone || '').trim();
  const relationship = (row.parent_relationship || '').trim() || null;

  if (!name || !phone) {
    return { contactId: null, status: 'skipped' };
  }

  // Check if contact exists for this user with this phone
  const existingResult = await client.query(
    'SELECT id FROM parent_contacts WHERE user_id = $1 AND phone_number = $2',
    [userId, phone]
  );
  const existing = existingResult.rows[0];

  if (existing) {
    // Update existing contact
    await client.query(
      `UPDATE parent_contacts
       SET name = $1,
           relationship = $2,
           updated_at = CURRENT_TIMESTAMP
       WHERE id = $3`,
      [name, relationship, existing.id]
    );
    return { contactId: existing.id, status: 'updated' };
  } else {
    // Insert new contact
    const contactId = uuidv4();
    await client.query(
      `INSERT INTO parent_contacts (id, user_id, name, phone_number, relationship)
       VALUES ($1, $2, $3, $4, $5)`,
      [contactId, userId, name, phone, relationship]
    );
    return { contactId, status: 'inserted' };
  }
}

async function processCSV(csvPath, dryRun = false) {
  const client = await pool.connect();
  const summary = {
    users_inserted: 0,
    users_updated: 0,
    contacts_inserted: 0,
    contacts_updated: 0,
    contacts_skipped: 0,
    rows_processed: 0,
    rows_failed: 0
  };

  try {
    // Read and parse CSV file
    const fileContent = fs.readFileSync(csvPath, 'utf-8');
    const records = parse(fileContent, {
      columns: true,
      skip_empty_lines: true,
      trim: true
    });

    await client.query('BEGIN');

    for (let idx = 0; idx < records.length; idx++) {
      const row = records[idx];
      try {
        // Upsert user
        const { userId, status: userStatus } = await upsertUser(client, row);
        summary[`users_${userStatus}`]++;

        // Upsert parent contact
        const { status: contactStatus } = await upsertParentContact(client, userId, row);
        summary[`contacts_${contactStatus}`]++;

        summary.rows_processed++;
      } catch (error) {
        summary.rows_failed++;
        console.error(`Row ${idx + 1} failed: ${error.message}`);
      }
    }

    if (dryRun) {
      await client.query('ROLLBACK');
      console.log('\nDry run complete; no changes committed.');
    } else {
      await client.query('COMMIT');
      console.log('\nImport committed to database.');
    }
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('Error processing CSV:', error);
    throw error;
  } finally {
    client.release();
  }

  return summary;
}

async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
    console.log('Usage: node bulk_import.js <csv_path> [--dry-run]');
    console.log('');
    console.log('Options:');
    console.log('  --dry-run    Validate without committing changes');
    process.exit(args.length === 0 ? 1 : 0);
  }

  const csvPath = args[0];
  const dryRun = args.includes('--dry-run');

  if (!fs.existsSync(csvPath)) {
    console.error(`Error: File not found: ${csvPath}`);
    process.exit(1);
  }

  try {
    const summary = await processCSV(csvPath, dryRun);
    
    console.log('\nImport summary:');
    for (const [key, value] of Object.entries(summary)) {
      console.log(`  ${key}: ${value}`);
    }
    
    await pool.end();
    process.exit(0);
  } catch (error) {
    console.error('Fatal error:', error.message);
    await pool.end();
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = { processCSV };
